### DISCUSSION TITLE
Discussion for the inclusion of a plugin API.

### DISCUSSION THREAD

**Discussion by AtriusX** [AtriusX/Waystones#22] [Ideas]
Recently there has been a bit of interest surrounding the need for integration with world protection plugins to ensure that waystones in the system cannot be misused (see PR #14). On top of this we should consider the ideas presented in Issues #19 and #20 as ideas beyond the scope of this project, but otherwise worth offering support for.

The purpose of this issue is to discuss what we should look for in an API. At this point it is clear we will need a standard one to expose access to some information about waystones within the system. It's also clear that an API will need to exist to offer first-class support to plugins like WorldGuard, and second-class support to other plugins.

What we need to consider is what both of these components will need and how we should implement them. Any ideas are welcome.

**Comment by GenericDevUsername** [comment]
There is an official worldguard API here 
https://worldguard.enginehub.org/en/latest/developer/index.html
As for other protection plugins maybe have a look at the more popular ones to see if they have an API too

**Reply by GenericDevUsername** [reply]
As for #19 and #20 :
#19 - For a plugin like this I don’t see a use for worldguard regions to be “discovered”. 
A better idea could be to allow for waystones placed in worldguard regions to automatically have that regions name  in their nametag when placed (toggles in config)

#20 - As a use I could see this as a cool concept to be added to the plugin, but from what I can tell the suggester is suggesting a GUI based warp system which isn’t really the goal of this plugin from what i gathered.
Maybe making an option for players to obtain admin warp keys that don’t have a use limit (if one time use is enabled in config) to warp to global waystones, (eg a waystones at a shopping district) These could be created by executing a command while holding a warp key. then distributed by third party /kit plugins or other means.
Another option would be to make any waystones with a certain non obtainable block under them to automatically create an admin warp key when clicked on with a compass.

**Reply by AtriusX** [reply]
The admin key suggestion is not a bad idea. I probably should point out that both of those suggestions would be possibly covered by the API, as I agree that both issues are out of scope for this plugin's intended purpose. The key is what sort of API we should offer for other plugins to make use of.

**Reply by GenericDevUsername** [reply]
An API would be a really good feature to implement, it's just a matter of pulling it off right so
It's understandable for developers.

**Reply by AtriusX** [reply]
As it stands right now, there are probably a few things the API could include such as:

- Ability to check if a location is a waystone
- Various properties and state of a waystone
- Retrieving all waystones within a given region or chunk (?)
- Whether a given player can use the waystone (useful for protection plugins)
- Events
    - Warp Initiate
    - Warp Cancel
    - Warp Success
    - Waystone Create
    - Waystone Destroy
    - Waystone Name
    - Waystone Info
- etc.

Any other ideas for this list are welcome.

**Reply by GenericDevUsername** [reply]
As of now I can't think of many other things the API would need to include.

I've been thinking of ideas while watching some Netflix and a couple of things have popped up.

1) the ability to limit the amount of waystones a player can place (could be waystones.limit.(number))

2) PlaceholderAPI placeholder support. A Couple of placeholders could be:
   - %waystones_Player_Placed% (amount of waystones placed by a player)
   - %Waystones_player_max% (max amount a player can place)

   - %Waystones_cooldown% (cooldown time in seconds)

   - %Waystones_Cooldown_Formatted% (cooldown time formatted [eg, 1m 30s])

**Reply by AtriusX** [reply]
In the case of an API, I don't think this is something that fits well with it since this would require the creation of a message service to use these formats. The ability for waystones to have a global cooldown is a pretty interesting idea though, as it stands cooldowns only apply on a per-player basis (via the waiting period). We can't really limit the number of waystones a player places though since we do not track that information within the plugin.

**Reply by GenericDevUsername** [reply]
I see, also I believe global cooldowns would need to be able to be toggled in config if people didn't want them.

**Reply by GenericDevUsername** [reply]
I'm gonna head to bed, we can continue this discussion tomorrow if you want.
One last question though. Do you happen to have a discord I can discuss this with you on? I check discord way more regularly than github so If you want to add me, my discord is CoronaCarrot# zero zero zero one

I typed the numbers because github removes the zeros otherwise... idk why

**Reply by AtriusX** [reply]
There would definitely be a config option for it. I would prefer to keep discussion within this thread though so a publicly visible record is available for anyone who might eventually be interested in the project.
